prov/gni/src/gnix_ep.c

```diff
@@ -393,11 +407,165 @@ int gnix_ep_open(struct fid_domain *domain, struct fi_info *info,
+		 * a local variable.
+		 */
+		atomic_inc(&gnix_id_counter);
+		fake_cdm_id = (domain_priv->cdm_id & 0x0000FFFF) |
+				((uint32_t)atomic_get(&gnix_id_counter) << 16);
+		status = GNI_CdmCreate(fake_cdm_id,
+					domain_priv->ptag,
+					domain_priv->cookie,
+					gnix_cdm_modes,
+					&nic->gni_cdm_hndl);
+		if (status != GNI_RC_SUCCESS) {
+			GNIX_LOG_ERROR("GNI_CdmCreate returned %s\n",
+					gni_err_str[status]);
+			/* TODO: need a translater from gni to fi errors */
+			ret = -FI_EACCES;
+			goto err;
+		}
+
+		/*
+		 * Okay, now go for the attach
+		*/
+		status = GNI_CdmAttach(nic->gni_cdm_hndl,
+					0,
+					&device_addr,
+					&nic->gni_nic_hndl);
+		if (status != GNI_RC_SUCCESS) {
+			GNIX_LOG_ERROR("GNI_CdmAttach returned %s\n",
+					gni_err_str[status]);
+			ret = -FI_EACCES;
+			goto err;
+		}
+
+		/*
+		 * create TX CQs - first polling, then blocking
+		 */
+
+		status = GNI_CqCreate(nic->gni_nic_hndl,
+					domain_priv->gni_tx_cq_size,
+					0,                    /* no delay count */
+					GNI_CQ_NOBLOCK |
+						domain_priv->gni_cq_modes,
+					NULL,                 /* useless handler */
+					NULL,                 /* useless handler context */
+					&nic->tx_cq);
+		if (status != GNI_RC_SUCCESS) {
+			ret = -FI_EINVAL;  /* TODO: better processing */
+			goto err;
+		}
+
+		status = GNI_CqCreate(nic->gni_nic_hndl,
+					domain_priv->gni_tx_cq_size,
+					0,                    /* no delay count */
+					GNI_CQ_BLOCKING |
+						domain_priv->gni_cq_modes,
+					NULL,                 /* useless handler */
+					NULL,                 /* useless handler context */
+					&nic->tx_cq_blk);
+		if (status != GNI_RC_SUCCESS) {
+			ret = -FI_EINVAL;  /* TODO: better processing */
+			goto err;
+		}
+
+		/*
+		 * create RX CQs - first polling, then blocking
+		 */
+
+		status = GNI_CqCreate(nic->gni_nic_hndl,
+					domain_priv->gni_rx_cq_size,
+					0,                    /* no delay count */
+					GNI_CQ_NOBLOCK |
+						domain_priv->gni_cq_modes,
+					NULL,                 /* useless handler */
+					NULL,                 /* useless handler context */
+					&nic->rx_cq);
+		if (status != GNI_RC_SUCCESS) {
+			ret = -FI_EINVAL;  /* TODO: better processing */
+			goto err;
+		}
+
+		status = GNI_CqCreate(nic->gni_nic_hndl,
+					domain_priv->gni_rx_cq_size,
+					0,                    /* no delay count */
+					GNI_CQ_BLOCKING |
+						domain_priv->gni_cq_modes,
+					NULL,                 /* useless handler */
+					NULL,                 /* useless handler context */
+					&nic->rx_cq_blk);
+		if (status != GNI_RC_SUCCESS) {
+			ret = -FI_EINVAL;  /* TODO: better processing */
+			goto err;
+		}
+
+		/*
+		 * TODO: set up work queue
+		 */
+
+		atomic_set(&nic->ref_cnt,1);
+		list_add_tail(&gnix_nic_list,&nic->list);
+	}
+
+	ep_priv->nic = nic;
 
 	atomic_inc(&domain_priv->ref_cnt);
-        *ep = &ep_priv->ep_fid;
-        return FI_SUCCESS;
+	*ep = &ep_priv->ep_fid;
+	return ret;
+
+err:
+	if (nic->rx_cq_blk != NULL) {
```

Comment:
`nic` may be NULL here due to goto on line 441, so must check `nic && nic->rx_cq` (similar for conditionals that follow).